chains/map_rerank_documents.go

```
@@ -0,0 +1,196 @@
+func NewMapRerankDocuments(mapRerankLLMChain *LLMChain) MapRerankDocuments {
+	mapRerankRE := `\s*(?P<answer>.*?)\nScore: (?P<score>.*)`
+	mapRerankLLMChain.OutputParser = outputparser.NewRegexParser(mapRerankRE)
+
+	return MapRerankDocuments{
+		LLMChain:                  mapRerankLLMChain,
+		MaxConcurrentWorkers:      1,
+		LLMChainInputVariableName: _combineDocumentsDefaultDocumentVariableName,
+		DocumentVariableName:      _combineDocumentsDefaultDocumentVariableName,
+		InputKey:                  _combineDocumentsDefaultInputKey,
+		OutputKey:                 _combineDocumentsDefaultOutputKey,
+		RankKey:                   _mapRerankDocumentsDefaultRankKey,
+		AnswerKey:                 _mapRerankDocumentsDefaultAnswerKey,
+	}
+}
+
+// Call handles the inner logic of the MapRerankDocuments chain.
+func (c MapRerankDocuments) Call(ctx context.Context, values map[string]any, options ...ChainCallOption) (map[string]any, error) { //nolint:lll
+	// Get the documents from the input key.
+	docs, ok := values[c.InputKey].([]schema.Document)
+
+	if !ok {
+		return nil, fmt.Errorf("%w: %w", ErrInvalidInputValues, ErrInputValuesWrongType)
+	}
+
+	if len(docs) == 0 {
+		return nil, fmt.Errorf("%w: documents slice has no elements", ErrInvalidInputValues)
+	}
+
+	applyInputs := c.getApplyInputs(values, docs)
+	mapResults, err := Apply(ctx, c.LLMChain, applyInputs, c.MaxConcurrentWorkers, options...)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a slice of outputs to return after sorting the ranks.
+	outputs := make([]map[string]any, len(mapResults))
+
+	for i, res := range mapResults {
+		rankedAnswer, ok := res[c.LLMChain.OutputKey].(map[string]string)
+		if !ok {
+			return nil, ErrInvalidOutputValues
+		}
+
+		outputs[i] = c.parseMapResults(rankedAnswer)
+	}
+
+	sort.Slice(outputs, func(i, j int) bool {
+		curr, err := strconv.Atoi(outputs[i][c.RankKey].(string))
+		if err != nil {
+			return false
+		}
```

Is there a reasoning for ignoring the error here? Why not do the integer conversion and sorting separately so we can handle conversion errors appropriately?

---

> Is there a reason for ignoring the error here?

Thought is that if it cannot be converted that we can assume the "compare" map should take precedence over this in sorting without needing to continue. With the inverse being true below.

> Why not do the integer conversion and sorting separately so we can handle conversion errors appropriately?

I had considered this, but I've been trying for parity with the python counterpart, which doesn't convert the type in the output:
```
[{'answer': ' Leo', 'score': '100'}]
```

That being said, there's still some cleanup and functionality to improve/fix, and I don't have a strong preference.

---

Can you point me to the piece of code you're talking about? 

Looks like the Python [source](https://github.com/hwchase17/langchain/blob/b96ac13f3d17b7b2c9295fe369ab81439c105aca/langchain/chains/combine_documents/map_rerank.py#L192) does attempt integer conversion and will throw an error if it can't perform the conversion.

---

Same source.

I'm fine on raising/returning on failed conversion, but noting the output isn't typed as an int:
```
(Pdb) sorted_res[0][0]['score']
'100'
(Pdb) type(sorted_res[0][0]['score'])
<class 'str'>
(Pdb) sorted_res[0][0]
{'answer': ' Leo', 'score': '100'}
```

---

Ok I think I see where the disconnect is now. I meant we should do the integer conversion of `score` for the sake of sorting, but the output can still have `score` as a string (which is the Python counterpart's behaviour)